src/zenml/orchestrators/step_run_utils.py

```diff
@@ -0,0 +1,466 @@
+        )
+
+        if not cache_enabled:
+            continue
+
+        if set(step.spec.upstream_steps) - finished_invocations:
+            continue
+
+        invocations.add(invocation_id)
+
+    return invocations
+
+
+def create_cached_step_runs(
+    deployment: "PipelineDeploymentResponse",
+    pipeline_run: PipelineRunResponse,
+    stack: "Stack",
+) -> Set[str]:
+    """Create all cached step runs for a pipeline run.
+
+    Args:
+        deployment: The deployment of the pipeline run.
+        pipeline_run: The pipeline run for which to create the step runs.
+        stack: The stack on which the pipeline run is happening.
+
+    Returns:
+        The invocation IDs of the created step runs.
+    """
+    cached_invocations: Set[str] = set()
+    visited_invocations: Set[str] = set()
+    request_factory = StepRunRequestFactory(
+        deployment=deployment, pipeline_run=pipeline_run, stack=stack
+    )
+
+    pipeline_model, pipeline_run = prepare_pipeline_run_model(
+        pipeline_run=pipeline_run
+    )
+
+    while (
+        cache_candidates := find_cacheable_invocation_candidates(
+            deployment=deployment,
+            finished_invocations=cached_invocations,
+        )
+        - visited_invocations
+    ):
+        for invocation_id in cache_candidates:
+            visited_invocations.add(invocation_id)
+
+            try:
+                step_run_request = request_factory.create_request(
+                    invocation_id
+                )
+                request_factory.populate_request(step_run_request)
+            except Exception as e:
+                # We failed to create/populate the step run. This might be due
+                # to some input resolution error, or an error importing the step
+                # source (there might be some missing dependencies). We just
+                # treat the step as not cached and let the orchestrator try
+                # again in a potentially different environment.
+                logger.debug(
+                    "Failed preparing step run `%s`: %s", invocation_id, str(e)
+                )
+                continue
+
+            if step_run_request.status != ExecutionStatus.CACHED:
+                # If we're not able to cache the step run, the orchestrator
+                # will run the step later which will create the step run
+                # -> We don't need to do anything here
+                continue
+
+            step_run = Client().zen_store.create_run_step(step_run_request)
+
+            # Refresh the pipeline run here to make sure we have the latest
+            # state
+            pipeline_run = Client().get_pipeline_run(pipeline_run.id)
+
+            step_model, step_run = prepare_step_run_model(
+                step_run=step_run, pipeline_run=pipeline_run
+            )
+
+            model = step_model or pipeline_model
+
+            # TODO: This function also imports the step code, which means it
```

Comment:
@avishniakov

Reply:
I would say the latest. Don't think that people are using that. Have no proofs though.